internal/encryption/keyprovider/aws_kms/provider.go

```diff
@@ -0,0 +1,98 @@
+package aws_kms
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/opentofu/opentofu/internal/encryption/keyprovider"
+)
+
+type keyMeta struct {
+	CiphertextBlob []byte `json:"ciphertext_blob"`
+}
+
+type keyProvider struct {
+	Config
+	svc *kms.Client
+	ctx context.Context
+}
+
+func (p keyProvider) Provide(rawMeta keyprovider.KeyMeta) (keyprovider.Output, keyprovider.KeyMeta, error) {
+	inMeta := rawMeta.(*keyMeta)
+	outMeta := keyMeta{}
+	out := keyprovider.Output{}
+
+	// Generate new key pair
+	if p.Symetric {
+		var spec types.DataKeySpec
+
+		for _, opt := range spec.Values() {
+			if string(opt) == p.KeySpec {
+				spec = opt
+			}
+		}
+
+		if len(spec) == 0 {
+			return out, outMeta, fmt.Errorf("Invalid key_spec %s, expected one of %v", p.KeySpec, spec.Values())
+		}
+
+		generatedKeyData, err := p.svc.GenerateDataKey(p.ctx, &kms.GenerateDataKeyInput{
+			KeyId:   aws.String(p.KMSKeyID),
+			KeySpec: spec,
+		})
+
+		if err != nil {
+			return out, outMeta, err
+		}
+
+		// Set inital outputs
+		out.EncryptionKey = generatedKeyData.Plaintext
+		out.DecryptionKey = generatedKeyData.Plaintext
```

Comment:
I'd add a comment here that the decryption key will be overriden later, if the statefile is already encrypted. Just because I was surprised when reading the code, and only later saw the end of the function.